File: bar_frontend/src/app/model/Cocktail.ts
import {User} from "./User";
import {Recipe} from "./Recipe";

export class Cocktail {
  cocktailId: number;
  cocktailName: string;
  cocktailAuthor: User;
  cocktailRating: number;
  publicationDate: Date;
  cocktailImage: string;
  cocktailImageThumbnail: string;
  cocktailRecipe: string;
  approxAlcoholPercentage: number;
  voteCount: number;
  recipes: Recipe[];


  constructor(cocktailId?: number, cocktailName?: string, cocktailAuthor?: User,
              cocktailRating?: number, publicationDate?: Date, cocktailImage?: string,
              cocktailImageThumbnail?: string, cocktailRecipe?: string, approxAlcoholPercentage?: number,
              voteCount?: number, recipes?: Recipe[]) {
    this.cocktailId = cocktailId;
    this.cocktailName = cocktailName;
    this.cocktailAuthor = cocktailAuthor;
    this.cocktailRating = cocktailRating;
    this.publicationDate = publicationDate;
    this.cocktailImage = cocktailImage;
    this.cocktailImageThumbnail = cocktailImageThumbnail;
    this.cocktailRecipe = cocktailRecipe;
    this.approxAlcoholPercentage = approxAlcoholPercentage;
    this.voteCount = voteCount;
    this.recipes = recipes;
  }


  // toJSON(): any {
  //   const { cocktailId, ...rest } = this;
  //   return {
  //     '@class': 'com.gbyzzz.bar_web_app.bar_backend.dto.CocktailDTO', // Manually include the JSON representation or null if cocktailImage is null/undefined
  //     ...rest
  //   };
  // }
}

File: bar_backend/src/main/java/com/gbyzzz/bar_web_app/bar_backend/entity/Cocktail.java
package com.gbyzzz.bar_web_app.bar_backend.entity;

import com.gbyzzz.bar_web_app.bar_backend.config.MinioConfig;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Formula;

import java.sql.Date;
import java.util.List;
import java.util.Set;

@Entity
@NoArgsConstructor
@Data
@Builder
@AllArgsConstructor
@EqualsAndHashCode
@Table(name = "cocktails", schema = "public", catalog = "bar_db")
public class Cocktail {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "cocktail_id")
    private Long cocktailId;

    @Basic
    @Column(name = "cocktail_name", length = 50, unique = true, nullable = false)
    private String cocktailName;

    @ManyToOne()
    @JoinColumn(name = "cocktail_author")
    private User cocktailAuthor;

    @Basic
    @Column(name = "cocktail_rating")
    private float cocktailRating;

    @Basic
    @Column(name = "publication_date")
    private Date publicationDate;

    @Basic
    @Column(name = "image_url")
    private String cocktailImage;

    @Basic
    @Column(name = "image_thumbnail_url")
    private String cocktailImageThumbnail;

    @Basic
    @Column(name = "cocktail_recipe")
    private String cocktailRecipe;

    @Basic
    @Column(name = "approx_alcohol_percentage")
    private int approxAlcoholPercentage;

    @OneToMany(mappedBy = "cocktail")
    private List<Recipe> recipes;

    @Formula(value = "(SELECT COUNT(*) FROM votes v WHERE v.cocktail_id=cocktail_id)")
    private int voteCount;

    @Override
    public String toString() {
        return "Cocktail{" +
                "cocktailId=" + cocktailId +
                ", cocktailName='" + cocktailName + '\'' +
                ", cocktailAuthor=" + cocktailAuthor +
                ", cocktailRating=" + cocktailRating +
                ", publicationDate=" + publicationDate +
                ", cocktailImage=" + cocktailImage +
                ", cocktailRecipe='" + cocktailRecipe + '\'' +
                ", approxAlcoholPercentage=" + approxAlcoholPercentage +
                ", voteCount=" + voteCount +
                '}';
    }
}
